### DISCUSSION TITLE
Tracking state of parallel states

### DISCUSSION THREAD

**Discussion by NixBiks** [statelyai/xstate#1909] [Q&A]
I've made a toy example in this [gist](https://xstate.js.org/viz/?gist=8c500c536b64f20a9892da7dc034aa7d). The idea is that my machine starts fetching data from different promises in parallel. 

The above works fine but requires the user of the machine to verify each state in the parallel state. I'd like to expose a method (or state) that tells whether or not all the parallel states are in `loaded`. Note I can't use `onDone` since then I leave the parallel state but I want to be able to refresh the data (without fetching from all services). Is it possible and how would that pattern look like?

**Comment by ezhikov** [comment]
You actually have that state already. Your state value composed of all active parallel states, so when you check with `matches` you can do this as such: `current.matches({ claims: "loaded", token: "loaded" })`.

Here is a [demo with your exact machine](https://codesandbox.io/s/frosty-lehmann-5dg0f?file=/src/index.js:2323-2377)

**Reply by NixBiks** [reply]
Yeah thanks for that. I ended up doing like [this](https://xstate.js.org/viz/?gist=a3e8fa6c5be8e0bb4aaf767dd437f303).

The idea is to create yet another parallel state that keeps track of whether my "other states" have been initialized since that's the state I'm interested in actually. I want to know whether I got data at least once or not and also whether the last attempt to refetch failed. Not in the gist above the following guard should be provided 

```javascript
notNullTokenAndClaims: (context, event) => !!context.claims && !!context.token
```

For some reason the visualizer breaks if I add that since `context` is undefined which I really don't understand. Must be a bug

**Reply by ezhikov** [reply]
With this solution you basically create more _possible_ states. With your initial machine you have 3 \* 3 (9) possible states. When you add another parallel state with 3 states, you get 3 \* 3 \* 3 (27) possible states, and some of them should be impossible, like `{claims: 'loaded', token: 'loaded', status: 'broken'}`. Best way to achieve absence of impossible states, in my experience, is to remove them from statechart.
